src/circuit/circuit_frame.rs

```diff
@@ -5722,13 +5778,88 @@ mod tests {
             )
             .unwrap();
 
-            popcount(
+            assert!(popcount(
                 &mut cs.namespace(|| format!("popcount {}", x.to_string())),
                 &bits,
                 &alloc_popcount.hash(),
-            );
+            )
+            .is_ok());
         }
 
         assert!(cs.is_satisfied());
     }
+
+    #[test]
+    fn test_to_u32() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+        let a = s.num(42);
+        let alloc_a = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "a"), s, || Ok(&a)).unwrap();
+        let add_pow_32 = add(
+            &mut cs.namespace(|| "add pow(2, 32)"),
+            &alloc_a.hash(),
+            &g.power2_32_num,
+        )
+        .unwrap();
+        let bits = add_pow_32.to_bits_le(&mut cs.namespace(|| "bits")).unwrap();
+        let v = match add_pow_32.get_value() {
+            Some(v) => v,
+            None => Fr::zero(), //dummy
+        };
+        let field_bn = BigUint::from_bytes_le(v.to_repr().as_ref());
+        let res =
+            to_unsigned_integer_helper(&mut cs, &g, &add_pow_32, field_bn, &bits, UnsignedInt::U32)
+                .unwrap();
+
+        let is_equal =
+            alloc_equal(&mut cs.namespace(|| "is equal"), &res, &alloc_a.hash()).unwrap();
+
+        assert!(enforce_true(&mut cs.namespace(|| "enforce true"), &is_equal).is_ok());
+        assert!(cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_to_u64() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+        let a = s.num(42);
+        let alloc_a = AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "a"), s, || Ok(&a)).unwrap();
```

Comment:
Why always using `AllocatedPtr` when you just an `AllocatedNum`? Sorry to keep pointing it out, but this happens throughout. It doesn't matter so much in tests, but it's confusing (and does matter outside of tests) — so I'd like to try to weed out the pattern.

Reply:
done